File: GVO Code v3/CalcCoord.pas
unit CalcCoord;

interface
uses utils, sysutils,windows,math;

Procedure UpdateCoord;

implementation
uses Main,GetQandYU,dialogs, OMS68SERMC ;

Procedure UpdateCoord;  {from stepper controller}
Var
 CmdStr,TempStr1,TempStr2,TempStr3 : String;
 x,y,alt,ha,Xcount,Ycount,TempStime : real;
 thandle : HWND;
Label DoDec;
//
//  read stepper counts
//  x = ra, y = dec
//
Begin {Update coord}    {from stepper controller}
   thandle := 0;
   MainForm.labeltkrate.Caption := floattostr(TrkRate);
   if(MainForm.PanelStatus.Visible = True) Then  MainForm.PanelStatus.Visible := False     //blink active indicator
      Else MainForm.PanelStatus.Visible := True;
//
    {Get Positions, wait for them}
    While (SendAndGetString(CommRecord,'PP;',Response)<> 0) Do ; //get position of all axis
       Begin
         TempList.CommaText := Response;
         x := strtoFloat(TempList[0]);                  // x = ra  counts
//         MainForm.Labelraval.Caption := TempList[0];   //write to Form     ra counts
         MainForm.labeldecval.Caption := TempList[1];   //write to Form    dec counts
       End;
//
    CmdStr := '';
    TempStime := GetStime;
    If(X = 0.0)Then X := 1.0;
//
// Ra Counts
// east of meridian is neg counts     (-ha)
// west of meridian is positive counts    (+ha)
//
    If( ((X > 0.0) AND (Ypole > 0)))Then              //west of meridian, above pole. quad 3
    Begin {quad 3}                              //positive counts
       RANow := (TempStime );                    //     meridian = 0 counts
       RANow := RANow - ((X/RAFact)/15.0) ;
       if(RANow < 0 ) Then RANow := RANow + 24;
//cal hr min, secs
       RAHr :=  Trunc(RANow);
       RAMin :=   Trunc(Frac(RANow) * 60);
       RASec :=   Trunc((Frac(RANow) * 3600)- (RAMin*60)) ;
       Str(RAHr:2,TempStr1) ;
// display ra
       MainForm.Rahredit.Text := TempStr1;
       Str(RAMin:2,TempStr2);
       MainForm.Raminedit.Text := TempStr2;
       Str(RAsec:2,TempStr3) ;
       MainForm.Rasecedit.Text := TempStr3;
       GoTo DoDec;
    End;{quad 3}
    If( (X < 0.0) AND (Ypole < 0) )Then
    Begin {Quad 2}
       X := Abs(X);
       RANow := (TempStime);
       RANow := RANow + ((X/RAFact)/15.0) ;
       if(RANow > 24 ) Then RANow := RANow - 24;
//calc hr, min, sec
       RAHr :=  Trunc(RANow);
       RAMin :=   Trunc(Frac(RANow) * 60);
       RASec :=   Trunc((Frac(RANow) * 3600)- (RAMin*60)) ;
// display ra
       Str(RAHr:2,TempStr1) ;
       MainForm.Rahredit.Text := TempStr1;
       Str(RAMin:2,TempStr2);
       MainForm.Raminedit.Text := TempStr2;
       Str(RAsec:2,TempStr3) ;
       MainForm.Rasecedit.Text := TempStr3;
       GoTo DoDec;
    End; {Quad 2}
    If( (X < 0.0) AND (Ypole > 0) )Then                    //east of meridian, above pole, quad 1
    Begin {Quad 1}
       X := Abs(X);
       RANow := (TempStime);
       RANow := RANow + ((X/RAFact)/15.0) ;
       if(RANow > 24 ) Then RANow := RANow - 24;
//calc hr, min, sec
       RAHr :=  Trunc(RANow);
       RAMin :=   Trunc(Frac(RANow) * 60);
       RASec :=   Trunc((Frac(RANow) * 3600)- (RAMin*60)) ;
// display ra
       Str(RAHr:2,TempStr1) ;
       MainForm.Rahredit.Text := TempStr1;
       Str(RAMin:2,TempStr2);
       MainForm.Raminedit.Text := TempStr2;
       Str(RAsec:2,TempStr3) ;
       MainForm.Rasecedit.Text := TempStr3;
       GoTo DoDec;
    End; {Quad 1}
    If( ((X > 0.0) AND (Ypole < 0)))Then
    Begin {quad 4}
       RANow := (TempStime );                    //     meridian = 0 counts
       RANow := RANow - ((X/RAFact)/15.0) ;
       if(RANow < 0 ) Then RANow := RANow + 24;
//cal hr min, secs
       RAHr :=  Trunc(RANow);
       RAMin :=   Trunc(Frac(RANow) * 60);
       RASec :=   Trunc((Frac(RANow) * 3600)- (RAMin*60)) ;
       Str(RAHr:2,TempStr1) ;
// display ra
       MainForm.Rahredit.Text := TempStr1;
       Str(RAMin:2,TempStr2);
       MainForm.Raminedit.Text := TempStr2;
       Str(RAsec:2,TempStr3) ;
       MainForm.Rasecedit.Text := TempStr3;
       GoTo DoDec;
    End;{quad 4}
//***********************{Do Dec Now} *************************************

//
//dec stepper counts positive when north of zenith, neg when south of zenith
//
DoDEc:
    CmdStr := '';
    Y := strtoFloat(TempList[1]);          //y = counts
    DECNow:= C_Lat  + (Y/DECFact);
//calc deg, min, sec
    decdeg :=  Trunc(decNow);
    decMin :=   Trunc(Frac(decNow) * 60);
    decSec :=   Trunc((Frac(decNow) * 3600)- (decMin*60)) ;
//display dec
    Str(decdeg:2,TempStr1) ;
    MainForm.decdegedit.Text := TempStr1;
    Str(decMin:2,TempStr2);
    MainForm.decminedit.Text := TempStr2;
    Str(decsec:2,TempStr3) ;
    MainForm.decsecedit.Text := TempStr3;
//**************{Set Quadrant Now}***********************************************
    Coord(pndomem^).RA := RANow; Coord(pndomem^).Dec := DecNow ;
    GetQandY(RANow,DecNow,Alt,HA,xcount,YCount,Quadrant,Ypole);
//check for below horizon (20 degrees)    
    if(alt <20) then
    begin
      cmdstr := 'AA ST;' ;
      SendString(CommRecord,pchar(CmdStr));
      MainForm.ButtonStop.Caption := 'START';
      Coord(pndomem^).RAGoTo := 0.0   ;
      Coord(pndomem^).decGoTo := 0.0  ;
      windows.MessageBox(thandle,'No track below horizon','Telescope Control',MB_SYSTEMMODAL);
    end;
//display alt, azimuth
    altdeg :=  Trunc(alt);
    altmin :=   Trunc(Frac(alt) * 60);
    altsec :=   Trunc((Frac(alt) * 3600)- (altmin*60)) ;
    Str(altdeg:2,TempStr1) ;
    MainForm.editaltdeg.Text := TempStr1;
    Str(altmin:2,TempStr2);
    MainForm.editaltmin.Text := TempStr2;
    Str(altsec:2,TempStr3) ;
    MainForm.editaltsec.Text := TempStr3;
//
    azimuth :=  (- sin(DECNow)) / cos(c_lat);
    if (azimuth < -1) then azimuth := -1;
    if (azimuth > 1) then azimuth := 1;
    azimuth := radtodeg(ArcCos( azimuth ));
    
    azdeg :=  Trunc(azimuth);
    azmin :=   Trunc(Frac(azimuth) * 60);
    azsec :=   Trunc((Frac(azimuth) * 3600)- (azmin*60)) ;
    Str(azdeg:2,TempStr1) ;
    MainForm.editazdeg.Text := TempStr1;
    Str(azmin:2,TempStr2);
    MainForm.editazmin.Text := TempStr2;
    Str(azsec:2,TempStr3) ;

    MainForm.Refresh;
{}
End; {Update Coord}
end.

File: GVO Code C++ v1/CalcCoord.cpp
#include "CalcCoord.h"
#include "GlobalValues.h"
#include "GetQandYU.h"
#include "CommUtils.h"
#include "utils.h"

#include <cmath>
#include <iostream>
#include <string>
#include <sstream>
#include <iomanip>

inline double radToDeg(double rad) { return rad * 180.0 / pi; }
inline double degToRad(double deg) { return deg * pi / 180.0; }

void UpdateCoord() {
	Coord* coord = CommUtils::GetCoordPtr();

	if (!coord) {
		return;
	}

	static bool last_pp_failed = false;

	int pp_result = SendAndGetCommand(&CommRecord, "PP;", Response, sizeof(Response));

	if (!pp_result) {
		if (!last_pp_failed) {
			std::cerr << "UpdateCoord: failed to query controller with PP; "
					  << "(current code treats return value 0 as failure)\n";
			last_pp_failed = true;
		}
		return;
	}

	last_pp_failed = false;

	std::string resp = Response;
	std::stringstream ss(resp);
	std::string xStr, yStr;

	// Split the comma-separated response exactly like Pascal TempList.CommaText usage
	bool gotX = static_cast<bool>(std::getline(ss, xStr, ','));
	bool gotY = static_cast<bool>(std::getline(ss, yStr, ','));

	double x = 0.0;
	double y = 0.0;

	try {
		x = std::stod(xStr); // RA counts
		y = std::stod(yStr); // DEC counts
	}
	catch (const std::exception& e) {
		std::cerr << "UpdateCoord: could not parse PP response: [" << resp << "]\n";
		std::cerr << "UpdateCoord: exception = " << e.what() << "\n";
		return;
	}
	catch (...) {
		std::cerr << "UpdateCoord: could not parse PP response: [" << resp << "]\n";
		std::cerr << "UpdateCoord: unknown parse exception\n";
		return;
	}

	double Alt = 0.0;
	double HA = 0.0;
	double Xcount = 0.0;
	double Ycount = 0.0;
	double TempStime = 0.0;

	TempStime = GetStime();

	if (x == 0.0) {
		std::cerr << "ERROR: RA encoder returning 0: invalid state\n";
		return;
	}

	double originalXForRA = x;

	if ((x > 0.0) && (yPole > 0)) {				// quad 3
		RANow = TempStime - ((x / RAFact) / 15.0);
		if (RANow < 0.0)
			RANow += 24.0;
	}
	else if ((x < 0.0) && (yPole < 0)) {		// quad 2
		x = std::abs(x);
		RANow = TempStime + ((x / RAFact) / 15.0);
		if (RANow > 24.0)
			RANow -= 24.0;
	}
	else if ((x < 0.0) && (yPole > 0)) {		// quad 1
		x = std::abs(x);
		RANow = TempStime + ((x / RAFact) / 15.0);
		if (RANow > 24.0)
			RANow -= 24.0;
	}
	else if ((x > 0.0) && (yPole < 0)) {		// quad 4
		RANow = TempStime - ((x / RAFact) / 15.0);
		if (RANow < 0.0)
			RANow += 24.0;
	}
	else {
		std::cout << "UpdateCoord: WARNING: no RA branch matched!"
				  << " originalX=" << originalXForRA
				  << " currentX=" << x
				  << " yPole=" << yPole
				  << "\n";
	}

	RAHr = static_cast<int>(RANow);
	RAMin = static_cast<int>((RANow - RAHr) * 60.0);
	RASec = static_cast<int>((RANow - RAHr) * 3600.0 - (RAMin * 60.0));

	DECNow = C_Lat + (y / DECFACT);

	DECDeg = static_cast<int>(DECNow);
	DECMin = static_cast<int>((DECNow - DECDeg) * 60.0);
	DECSec = static_cast<int>((DECNow - DECDeg) * 3600.0 - (DECMin * 60.0));

	coord->RA = RANow;
	coord->Dec = DECNow;

	GetQandY(RANow, DECNow, Alt, HA, Xcount, Ycount, quadrant, yPole);

	// Safety stop below 30 degrees altitude, same as Pascal behavior
	if (Alt < 30.0) {
		SendCommand("AA ST;");   // stop all axes

		coord->RAGoto = 0.0;
		coord->DecGoto = 0.0;

		movingRA = false;
		movingDEC = false;

		std::cerr << "No track below 30 degrees!\n";
		lowAltitudeLockout = true;
		return;
	}
	
	altdeg = static_cast<int>(Alt);
	altmin = static_cast<int>((Alt - altdeg) * 60.0);
	altsec = static_cast<int>((Alt - altdeg) * 3600.0 - (altmin * 60.0));

	double azimuthArg = (-std::sin(degToRad(DECNow))) / std::cos(degToRad(C_Lat));

	if (azimuthArg < -1.0) azimuthArg = -1.0;
	if (azimuthArg > 1.0) azimuthArg = 1.0;

	double azimuth = radToDeg(std::acos(azimuthArg));

	azdeg = static_cast<int>(azimuth);
	azmin = static_cast<int>((azimuth - azdeg) * 60.0);
	azsec = static_cast<int>((azimuth - azdeg) * 3600.0 - (azmin * 60.0));
}